conan_provider.cmake

```diff
@@ -19,18 +17,53 @@ function(detect_os OS OS_API_LEVEL OS_VERSION)
             message(STATUS "CMake-Conan: android_platform=${ANDROID_PLATFORM}")
             set(${OS_API_LEVEL} ${_OS_API_LEVEL} PARENT_SCOPE)
         endif()
+        if(${CMAKE_SYSTEM_NAME} STREQUAL "Darwin" OR
+           ${CMAKE_SYSTEM_NAME} STREQUAL "iOS" OR
+           ${CMAKE_SYSTEM_NAME} STREQUAL "tvOS" OR
+           ${CMAKE_SYSTEM_NAME} STREQUAL "watchOS")
+            if(DEFINED CMAKE_OSX_SYSROOT)
+                # CMAKE_OSX_SYSROOT contains the full path to the SDK for MakeFile/Ninja
+                # generators, but just has the original input string for Xcode. Use the
+                # lowercase form to detect both.
+                string(TOLOWER ${CMAKE_OSX_SYSROOT} cmake_os_sysroot_lower)
+                if(cmake_os_sysroot_lower MATCHES appletvos)
+                    set(_OS_SDK appletvos)
+                elseif(cmake_os_sysroot_lower MATCHES appletvsimulator)
+                    set(_OS_SDK appletvsimulator)
+                elseif(cmake_os_sysroot_lower MATCHES iphoneos)
+                    set(_OS_SDK iphoneos)
+                elseif(cmake_os_sysroot_lower MATCHES iphonesimulator)
+                    set(_OS_SDK iphonesimulator)
+                elseif(cmake_os_sysroot_lower MATCHES watchos)
+                    set(_OS_SDK watchos)
+                elseif(cmake_os_sysroot_lower MATCHES watchsimulator)
+                    set(_OS_SDK watchsimulator)
+                endif()
+                message(STATUS "CMake-Conan: cmake_osx_sysroot=${CMAKE_OSX_SYSROOT}")
+                set(${OS_SDK} ${_OS_SDK} PARENT_SCOPE)
+            endif()
+            if(DEFINED CMAKE_OSX_DEPLOYMENT_TARGET)
+                message(STATUS "CMake-Conan: cmake_osx_deployment_target=${CMAKE_OSX_DEPLOYMENT_TARGET}")
+                set(${OS_VERSION} ${CMAKE_OSX_DEPLOYMENT_TARGET} PARENT_SCOPE)
+            endif()
+        endif()
     endif()
 endfunction()
 
 
 function(detect_arch ARCH)
-    if(CMAKE_SYSTEM_PROCESSOR STREQUAL aarch64)
+    # CMAKE_OSX_ARCHITECTURES can contain multiple architectures, but Conan only supports one.
+    # Therefore this code only finds one. If the recipes support mulitple architectures, the
+    # build will work. Otherwise, there will be a linker error for the missing architecture(s).
```

Comment:
based on this comment and the logic, I'm not sure this is worth leaving here. If `CMAKE_OSX_ARCHITECTURES` is defined, then we cannot map to a single architecture. If I'm reading the conditional properly, with the proposed changes in this PR the conan architecture would be whichever matches `CMAKE_SYSTEM_PROCESSOR` or `CMAKE_OSX_ARCHITECTURES` when going in order (arm64, armv7-a, armv7, armv7s, x86_64) - not sure this makes for a predictable experience especially considering that unless the caller has made provisions to ensure that the binaries are fat binaries, there will be linker errors.

I would suggest falling back to issuing a warning when `CMAKE_OSX_ARCHITECTURES` is defined, and stick to following what `CMAKE_SYSTEM_PROCESSOR` says. In the future we can evaluate how to best handle this case but I have concerns this is fragile at the moment.

Reply:
Thanks @ssrobins!! Where would `CMAKE_OSX_ARCHITECTURES` be initialized? Does CMake give it default values if it is blank?
As for `CMAKE_SYSTEM_PROCESSOR`, this is typically set in the same toolchain file that also sets `CMAKE_SYSTEM_NAME` to `iOS` - in fact when cross-building this is one of the recommended variables to set in a toolchain, so I'm wondering what the most conventional way is to configure an iOS/tvOS/watchOS build with CMake? I think when Xcode is the generator possibly Cmake can configure the build without an architecture as one can be selected at build time instead with Xcode?

Reply:
`CMAKE_OSX_ARCHITECTURES` would be set by the user, as [shown in the documentation](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#cross-compiling-for-ios-tvos-or-watchos), which I would consider the conventional way to build iOS/tvOS/watchOS with CMake since [it was added in 3.14](https://cmake.org/cmake/help/latest/release/3.14.html#platforms). Before then, we had to rely on [community-supported toolchains](https://github.com/leetal/ios-cmake/blob/master/ios.toolchain.cmake), which did set `CMAKE_SYSTEM_PROCESSOR`.

`CMAKE_OSX_ARCHITECTURES` is blank by default, but the generated build defaults to the build machine's architecture so in that case `CMAKE_SYSTEM_PROCESSOR` is helpful.

Reply:
Added a warning if `CMAKE_OSX_ARCHITECTURES` contains more than one architecture.